### ISSUE TITLE
NonValidActionsError when training HyperGrid environment

### ISSUE THREAD

**Issue by JosephRRB** [saleml/torchgfn#51]
Hello everyone,

Thank you for making this repo. My colleagues and I are applying GFlowNets on a hypergrid environment using our own reward distribution. However upon training, we get the _NonValidActionsError_. I was able to reproduce the error using the built-in hypergrid environment but I can't isolate it. I can also provide the jupyter notebook where I got the error, if needed. The code to reproduce it is below:

```python
import torch

from gfn.envs import HyperGrid
from gfn import LogitPBEstimator, LogitPFEstimator, LogZEstimator
from gfn.losses import TBParametrization, TrajectoryBalance
from gfn.samplers import DiscreteActionsSampler, TrajectoriesSampler


torch.manual_seed(0)

exploration_rate = 0.5
learning_rate=0.0005

env = HyperGrid(ndim=5, height=2)

logit_PF = LogitPFEstimator(
    env=env,
    module_name="NeuralNet",
)
logit_PB = LogitPBEstimator(
    env=env,
    module_name="NeuralNet",
    torso=logit_PF.module.torso,
)
logZ = LogZEstimator(torch.tensor(0.0))

training_sampler = TrajectoriesSampler(
    env=env,
    actions_sampler=DiscreteActionsSampler(
        estimator=logit_PF,
        epsilon=exploration_rate
    )
)

parametrization = TBParametrization(logit_PF, logit_PB, logZ)
loss_fn = TrajectoryBalance(
    parametrization=parametrization,
)

params = [
    {
        "params": [
            val for key, val in parametrization.parameters.items() if "logZ" not in key
        ],
        "lr": learning_rate,
    },
    {"params": [val for key, val in parametrization.parameters.items() if "logZ" in key], "lr": 0.1},
]
optimizer = torch.optim.Adam(params=params)
``` 
```python
from tqdm import tqdm


n_iterations = int(1e4)
batch_size = int(1e5)

for i in (pbar := tqdm(range(n_iterations))):
    trajectories = training_sampler.sample(
        n_trajectories=batch_size
    )

    optimizer.zero_grad()
    loss = loss_fn(trajectories)
    loss.backward()
    optimizer.step()

    if i % int(1e4) == 0:
        pbar.set_postfix({"loss": loss.item()})
``` 

I followed the instructions to install _torchgfn_ and I got the torch version `2.0.1+cu117`.

The _DiscreteActionsSampler_ seems produce an invalid action at a random time when running the whole code but I can't see the error when I'm using a debugger to run the code line by line at the appropriate iteration. 

Please let us know what can be done to resolve this. 

Thank you very much and have a nice day!

**Comment by saleml** [comment]
Hi Joseph,

I am unable to reproduce your error, even when starting from an empty conda environment and doing `pip install torchgfn[scripts]`. However, note that I did not run your script _exactly_, but I changed the batch size to 100 rather than 100,000 (it was too slow). Are you really using a batch size of 1e5 ?

The `NonValidActionsError` is thrown in this line https://github.com/saleml/torchgfn/blob/master/src/gfn/envs/env.py#L175. Maybe you could add a breakpoint there ?

**Comment by JosephRRB** [comment]
Hello Salem,

Thank you for your reply! 

Indeed, the batch size of 100 did not reproduce the error. But I used the 1e5 batch size so that the error appears earlier in the training (the thought was the algorithm will see more states and would encounter the error faster). Specifically for the 1e5 batch size, the error occurs at the 40th iteration (according to tqdm). 

I added the breakpoint as requested and I got: 
`valid_states_masks[22939, :]` --> `tensor([ True,  True,  True, False,  True,  True])` 
`valid_actions[22939]` --> `tensor(3)`

So there seems to be an invalid action that slips through?

We do have other hypergrid environment cases like `(ndim=5, height=13)` and the error appeared much later in the training.

Can you try again with the 1e5 batch size? I did `pip install torchgfn[scripts]` on a newly created conda environment and I'm still getting the error. Let me try with a smaller batch size as well and I'll report back.

**Comment by JosephRRB** [comment]
Setting `batch_size = int(1e3)` raises the `NonValidActionsError` at the iteration `1874` (according to tqdm). Can you try the batch size of 1000?

**Comment by saleml** [comment]
With both 1e5 and 1e3, my code doesn't break at the iterations you specified; which makes me wonder why we have different behaviors. Are you using python=3.11 ?

Could you maybe replace the last two lines with:
```
if i % int(1e2) == 0:
        pbar.set_postfix({"loss": loss.item()})
        print(loss.item())
```
rerun with batch_size = int(1e3), then paste the ~100 (or less) lines from your terminal here ?

**Comment by JosephRRB** [comment]
If needed, would it be possible to share my screen with you reproducing the error?

**Comment by saleml** [comment]
It looks like a pytorch error. Here is why I think so:
If I debug the `sample` method of `actions_sampler.py` at the step at which the error is raised, something unexpected happens:
- `states` is HyperGridStates object of batch shape (255,) and state shape (5,)
- `probs` is a tensor of shape `torch.Size([255, 6])`
- The 127th entry of `probs` is `probs[127] = tensor([0.0000, 0.0000, 0.3646, 0.0000, 0.0000, 0.6354], grad_fn=<SelectBackward0>)`
- Yet, even though `probs` is used to define `dist = Categorical(probs=probs)`, which in turn is used to produce actions via `actions = dist.sample()`, we get `actions[127] = 1`.

It does seem like the Categorical distribution samples an index with zero probability ! Basically, the 127th state is `tensor([1, 1, 0, 1, 1])` so it makes sense that the only possible actions should be `2` (increase the 3rd dimension by 1) or `5` (exit), which is well encoded in `probs[127]`. The sampled action `1` is indeed invalid at this state. **It should not have been sampled**.

I found two issues in pytorch reporting a similar behavior: https://github.com/pytorch/pytorch/issues/100884 and https://github.com/pytorch/pytorch/issues/91863.

I would rather not make any change to this repo, as long as it's not solved from pytorch's side. 

In the meantime, I suggest replacing [the line `actions = dist.sample()`](https://github.com/saleml/torchgfn/blob/master/src/gfn/samplers/actions_samplers.py#L122) with:
```
while True:
    actions = dist.sample()
    if torch.all(torch.gather(states.forward_masks, 1, actions.unsqueeze(-1)).squeeze()).item():
        break
```

Would that be possible for your project ?

**Comment by JosephRRB** [comment]
Thanks for the fix! It didn't break now. So basically it would resample whenever it detects an invalid action, right?

Would it be possible to create a branch with this fix so that we can install directly from that branch? What do you think?

**Comment by saleml** [comment]
> Thanks for the fix! It didn't break now. So basically it would resample whenever it detects an invalid action, right?

Exactly

> Would it be possible to create a branch with this fix so that we can install directly from that branch? What do you think?

Sure.

You can now install via:
```
git clone https://github.com/saleml/torchgfn.git
conda create -n gfn python=3.11
conda activate torchgfn
cd torchgfn
git checkout fixsampling
pip install .
```

**Comment by JosephRRB** [comment]
Thank you very much again! I've changed our requirements.txt to point to the `fixsampling` branch now and we can continue our study.

**Comment by josephdviviano** [comment]
I just wanted to follow up with this -- is it still an issue?

I'm running the following on device without any errors.

```
import torch
from tqdm import tqdm

from gfn.gym import HyperGrid
from gfn.gflownet import TBGFlowNet
from gfn.modules import DiscretePolicyEstimator
from gfn.samplers import Sampler
from gfn.utils import NeuralNet


torch.manual_seed(0)
exploration_rate = 0.5
learning_rate = 0.0005

# Setup the Environment.
env = HyperGrid(
    ndim=5,
    height=2,
    device_str="cuda" if torch.cuda.is_available() else "cpu",
)

# Build the GFlowNet.
module_PF = NeuralNet(
    input_dim=env.preprocessor.output_dim,
    output_dim=env.n_actions,
)
module_PB = NeuralNet(
    input_dim=env.preprocessor.output_dim,
    output_dim=env.n_actions - 1,
    torso=module_PF.torso,
)
pf_estimator = DiscretePolicyEstimator(
    module_PF, env.n_actions, is_backward=False, preprocessor=env.preprocessor
)
pb_estimator = DiscretePolicyEstimator(
    module_PB, env.n_actions, is_backward=True, preprocessor=env.preprocessor
)
gflownet = TBGFlowNet(init_logZ=0.0, pf=pf_estimator, pb=pb_estimator, off_policy=True)

# Feed pf to the sampler.
sampler = Sampler(estimator=pf_estimator)

# Move the gflownet to the GPU.
if torch.cuda.is_available():
    gflownet = gflownet.to("cuda")

# Policy parameters have their own LR. Log Z gets dedicated learning rate
# (typically higher).
optimizer = torch.optim.Adam(gflownet.pf_pb_parameters(), lr=1e-3)
optimizer.add_param_group({"params": gflownet.logz_parameters(), "lr": 1e-1})

n_iterations = int(1e4)
batch_size = int(1e5)

for i in (pbar := tqdm(range(n_iterations))):
    trajectories = sampler.sample_trajectories(
        env,
        n_trajectories=batch_size,
        off_policy=True,
        epsilon=exploration_rate,
    )
    optimizer.zero_grad()
    loss = gflownet.loss(env, trajectories)
    loss.backward()
    optimizer.step()
    pbar.set_postfix({"loss": loss.item()})

```

with the following build:

```
In [1]: import torch

In [2]: torch.__version__
Out[2]: '2.0.1+cu117'
```

**Comment by josephdviviano** [comment]
The above example can now be found in this PR https://github.com/GFNOrg/torchgfn/pull/159

`train_hypergrid_simple.py`

**Comment by saleml** [comment]
> pytorch/pytorch#91863

Hi @josephdviviano 

Please have a look at my comment here: https://github.com/GFNOrg/torchgfn/issues/51#issuecomment-1579252978. 
It looks like the two pytorch issues remain open. Basically, this issue is due to pytorch's Categorical distribution sampling an element with 0 probability. I wonder how #159 solves it

**Comment by saleml** [comment]
Right! I suggest we leave the issue open until the pytorch bug is fixed, just in case someone else faces the same problem

**Comment by josephdviviano** [comment]
> It looks like a pytorch error. Here is why I think so: If I debug the `sample` method of `actions_sampler.py` at the step at which the error is raised, something unexpected happens:
> ...
> I found two issues in pytorch reporting a similar behavior: [pytorch/pytorch#100884](https://github.com/pytorch/pytorch/issues/100884) and [pytorch/pytorch#91863](https://github.com/pytorch/pytorch/issues/91863).
> 
> I would rather not make any change to this repo, as long as it's not solved from pytorch's side.
> 
> In the meantime, I suggest replacing [the line `actions = dist.sample()`](https://github.com/saleml/torchgfn/blob/master/src/gfn/samplers/actions_samplers.py#L122) with:
> 
> ```
> while True:
>     actions = dist.sample()
>     if torch.all(torch.gather(states.forward_masks, 1, actions.unsqueeze(-1)).squeeze()).item():
>         break
> ```
> 
> Would that be possible for your project ?

It looks like https://github.com/pytorch/pytorch/issues/100884 is closed, but not https://github.com/pytorch/pytorch/issues/91863.

@JosephRRB - is this behaviour still an issue for you?  I wonder if it has been resolved on the pytorch side and they haven't updated their issue tracker properly.

**Comment by JosephRRB** [comment]
Hello @josephdviviano, thank you for following up! The fix you've quoted was enough for our purposes. I'm uncertain about the pytorch level fixes that they have. I'd be more than happy to close the issue only our project was being affected by this,

**Comment by josephdviviano** [comment]
I'd like to test it out, because others would be affected by the bug as well in theory (and it would be very hard to detect).
